tree/tree/src/InternalTreeUtils.cxx

```diff
@@ -446,38 +460,65 @@ std::vector<std::string> ExpandGlob(const std::string &glob)
    delete[] epath;
 
    if (dir) {
-      // Create a TList to store the file names (not yet sorted)
-      TList l;
-      l.SetOwner(); // Make sure the TList will delete its objects
       TRegexp re(basename.c_str(), true);
       const char *file;
       TString fname;
+
       while ((file = gSystem->GetDirEntry(dir))) {
-         if (!strcmp(file, ".") || !strcmp(file, ".."))
+         if (!strcmp(file, ".") || !strcmp(file, "..")) {
             continue;
+         }
          fname = file;
-         if ((basename != file) && fname.Index(re) == kNPOS)
+
+         if ((basename != file) && fname.Index(re) == kNPOS) {
             continue;
-         // Using '/' as separator here as it was done in TChain::Add
-         // In principle this should be using the appropriate platform separator
-         l.Add(new TObjString((dirname + '/' + file).c_str()));
+         }
+
+         bool isDirectory = std::filesystem::is_directory(std::filesystem::path(dirname + "/" + file));
+         if (!remainder.empty() && isDirectory) {
+            RecursiveGlob(l, dirname + '/' + file + "/" + remainder);
```

Comment:
To make this line easier to read, consider renaming the variable `file` into `direntry` as it can be the name either a file or a directory.